signaling: signaling server and messaging over WebRTC

shalins opened the pull request:
### Update (7/9): 

Done with v1! `Communicator` is an incredibly simple interface that should easily let you establish a WebRTC connection and send any object from one peer to another.

---

Note: This is WIP! Feel free to look through this, but keep in mind this will go through major restructuring over the next few days. Everything is in an experiments folder as a temporary sandbox. This will be integrated into our main project.

### Description of Changes
- Created a simple experimental react client that handles initiating a peer connection over WebRTC
- The react client can also create serialized message payloads
- Create simple node express server to act as a signaling server.
- This signaling server can receive and deserialize the message payload that come from the client.
- Integrated websocket to establish connection between the clients the signaling server before direct RTC connection is established.

kanyesthaker commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+     * @param {CommunicatorCallback} callback The callback to call when the
+     * event is fired.
+     */
+  setEventCallback(event: string, callback: CommunicatorCallback) {
```
are these callbacks defined by the client?

kanyesthaker commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+  sendWebsocket(data: any) {
+    if (!this.isSocketConnected) {
+      if (this.onErrorCallback) {
+        this.onErrorCallback(new Error('Error: Not connected to the server.'));
```
style nit: `err = new Error(...); this.onErrorCallback(err);`

kanyesthaker commented on src/network/utils.ts:
```diff
@@ -0,0 +1,86 @@
+/* eslint-disable no-unused-vars */
```
duped file?

kanyesthaker replied:
i guess not duped, the two utils files are not exactly the same

kanyesthaker commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+    );
+
+    // Save this information locally.
+    save('peer', peer);
```
this key is non-unique right? so if we get multiple `join`s this will get overwritten?

shalins replied:
A `JOIN` message is sent from the Peer to the Signaling Server when the peer has no ID yet and has not joined any room. Then the server sends the peer it's _unique_ `peerID` (which is simply a `UUID`) and helps the peer join it's intended room. 

This process happens when the peer either (a) joins for the first time ever, or (b) closes the tab and then reopens the page. In case (a), we want to request a new a peerID from the signaling server so that we can keep track of that internally. So when the signaling server finally sends us our ID, we want to `save` it locally no matter what.

In case (b), we want to load our `peerID` from our cache, and send it to the signaling server as part of our `JOIN` message to basically tell the server "Hey, I've already joined before, and here was the ID you gave me!" Then the server can decide what to do. If it finds your information, it can just resume your session send you that same ID back. OR, If it realizes that the room you were in was deleted (since you were the only peer in it and you left), then it may have deleted the room. In which case it will send you a NEW `peerID`. In both cases, if you get a `peerID` as part of the server `JOIN` response. You should treat that as the source of truth and save it.

So overwriting the peerID should not be a problem.

kanyesthaker commented on tsconfig.json:
```diff
@@ -1,5 +1,6 @@
 {
   "compilerOptions": {
+    "target": "es5",
```
es6?

shalins replied:
Yup, fixed.

kanyesthaker reviewed (commented):
Good PR. Well written, thoughtful, and well documented -- I could get a good core understanding of what's going on even without much background in the mechanics of WebRTC. Left some comments, mostly around code style and organization, along with some logic errors/edge cases, along with some questions about design decisions. I'll do a second review later when I have a change to play around with the actual functionality, but great job overall!

samjgorman reviewed (commented):
Great work @shalins! Lots of positives to highlight-- such as many well-documented functions, clean layers of abstraction, and non-trivial functionality. 

My comments are a mix of questions on broader design and style. 

I found it hard to test for correctness in code review.  I would advocate for setting up some level of lightweight testing for less obvious parts of the signaling server to verify correctness and major edge cases. This could fit into the immediate plans of Spacebar well, where @kanyesthaker and I have already found several tangible use cases where a testing library such as Jest could speed up development in the long run.  

I'd really like to avoid this large PR from growing in scope, so this is a non-blocking comment. I'd be happy to collaborate on setting up a simple Jest testing environment and writing unit tests for major parts of our network logic and signaling server in a next PR.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,229 @@
+
+    let peer: Peer;
+
+    if (peersInRoom.length !== 0 && cachedPeer) {
```
This was done intentionally, since generally I reserve the `if (someVariable) { ... }` pattern for nullity checks since it prevents obfuscation to the reader of the code. 

The way `if peersInRoom.length !== 0` reads in English to me is `if the number of peers in the room is not zero`, whereas something like `if peersInRoom.length` reads closer to `if the peersInRoom.length object exists`, which are semantically slightly different things. 

In this case, I want the reader of this code to be able to translate the condition to mean `if the number of peers in the room is not zero and the cachedPeers object exists`.

shalins commented on src/network/communicator.ts:
```diff
@@ -0,0 +1,75 @@
+  // TODO(SHALIN): Deploy to Heroku to get something like wss://spacebar-ws-2.herokuapp.com/
+  private static socket_url = 'localhost';
+
+  private static socket_port = 3400;
```
Good catch! Fixed.

shalins commented on src/network/utils.ts:
```diff
@@ -0,0 +1,86 @@
+ * @return {any} The deserialized data object that was retreived from
+ * storage, or empty dictionary if object was not found.
+ */
+export function load(key: string) {
```
Yeah, great point! Updated.

shalins commented on src/network/utils.ts:
```diff
@@ -0,0 +1,86 @@
+ * @param {any} byteArray Byte array containing the serialized JSON object.
+ * @return {any} The deserialized payload object.
+ */
+export function parse(byteArray: any) {
```
Yeah that's a fair point. Fixed!

shalins commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+   * @param {any} event The event object containing the message
+   * sent by the signaling server.
+   */
+  _handleWebsocketReceive(event: any) {
```
Great point! I think I'll leave this out of scope for this PR since it is a bit of a major change, yes ideally all the JSON parse's should be converted to types with proper error handling (https://stackoverflow.com/questions/38688822/how-to-parse-json-string-in-typescript). For now, I've fixed some of the `any` types to better types, but will do this properly soon.

shalins commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+    // TODO(SHALIN): Verify that URL and PORT are valid.
+    this.ws = new WebSocket(`ws://${socketUrl}:${port}`);
+
+    this.ws.onopen = (event) => {
```
I agree. Unfortunately, these variables were not defined by me -- I'm just assigning functions to the pre-defined `Event` properties defined in the WebSocket API: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket

shalins commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+  constructor(socketUrl: string, port: number, roomID: string) {
+    this.peerManager = new PeerManager(
+      // Bind the current `this` to the callbacks so it refers to
+      // the SignalingClient instance.
```
Fixed!

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+    log.debug(`Join Request Received with Room ID: ${roomID}`);
+
+    // Check if the room exists. If not, create it.
+    if (!(roomID in this.rooms)) {
```
I'm not quite sure what you mean by unknown ID (unknown `peerID` or unknown `roomID`?). 

Currently the way the API is set up, when the Peer initializes the Communicator class, the peer MUST specify a roomID (because otherwise how would the signaling server know which set of other peers to connect the peer to?)

This makes it very easy to do something like link-based room creation, e.g. peer1 goes to `spacebar.com/xyz` and room `/xyz` is created. Then anyone else who goes to `/xyz` would be established a P2P connection with all others in that room.

This also means the frontend or client of Communicator can handle how they want to specify the room ID (asking the user to name their room, autogenerating a roomID, abstracting it away from the user, etc). As long as the two peers share that roomID, they will be matched.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+      this.rooms[roomID].peers.push(peer);
+    }
+
+    // Cache the websocket object this peer is connected over so we can
```
Fair point, changed to `Store` instead of `Cache`.

shalins commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+    );
+
+    // Save this information locally.
+    save('peer', peer);
```
Yup, done!

shalins commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,301 @@
+    this.peerManager.isInitiator = peer.initiator;
+
+    log.debug(
+      'We just received our id ${peer.peerID} and we are ',
```
Great suggestion -- fixed.

shalins commented on signaler/utils.ts:
```diff
@@ -0,0 +1,60 @@
+ * intends to join or is a part of.
+ * @return {any} A serialized JSON string.
+ */
+export function createPayload(
```
Why? If there is a logical reason for doing so, or an idiomatic style convention in Typescript that I'm not aware of, do you mind sharing a source? It would be helpful for me to understand why such a convention is maintained.

As far as my understanding & research on the topic goes, Javascript arrow functions are similar to what in other languages is referred to as closures (RUST, Swift, etc.).

I can see the argument for using arrow functions in cases where you don't need a function to be named and anonymous functions will both clean up the readability, as well as allow you to properly scope `this` without needing to use `.bind()`.

For example:

```
exampleFunction() {
  setTimeout((function() { console.log(`My name is ${this.name}`) }).bind(this), 10)
}
```

to this:
```
exampleFunction() {
  setTimeout(() => { console.log(`My name is ${this.name}`) }, 10)
}
```

In my case, it seems as though using a traditional function is perfectly okay, and and I'd go as far as to say it's probably better since it's pretty obvious to read and understand what's going on vs using an anonymous arrow function, then binding it to a variable. 

Let me know what you think or if I'm mistaken, I'm happy to revisit!

Sources: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

kanyesthaker replied:
bumping this per our discussion yesterday -- whether it's worth using `function` vs arrow notation for functions that don't use `this`; whatever decision we come to should be consistent through the project; (fwiw: from a scan of the internet, most javascript libraries that use es6 arrow notation tend to stick with arrow notation, and most that use es5 function notation tend to stick with that)

shalins replied:
Agree with the point about consistency, fixed!

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+     * @param {string}message Message to send.
+     */
+  broadcast(roomID: string, peerID: string, message: string) {
+    this.rooms[roomID].peers.forEach((peer) => {
```
I thought about this -- but I'd say right now, no. Whereas SocketIO has inbuilt functionality for maintaining rooms, active peer connections, etc. Websocket does not. So I'd have to write the logic to check for active vs stale connections myself. Since this doesn't super matter for connecting 2-3 peers, I'm going to defer this to when it does start mattering. Great question though.

shalins commented on signaler/utils.ts:
```diff
@@ -0,0 +1,60 @@
+export enum PayloadType {
```
Great point. Currently there were two things done intentionally. 

1. I have mostly duplicated the utility file in both the Signaling Server as well as the Communicator API. Ideally this would be merged into one file. However, for now I kept it in separate files so that I can deploy the server to Heroku and make changes easily without it affecting the communicator API. 
2. I've also put everything in one file even though you're right, the PayloadType belongs in another file, maybe something like `types.ts`. For now, I will keep it in one file until I successfully deploy the server code to Heroku, and at that point I can make a small PR splitting this up into better modules and also using symlinks instead of duplicate files.

shalins commented on src/network/communicator.ts:
```diff
@@ -0,0 +1,75 @@
+/* eslint-disable no-unused-vars */
```
Good catch -- removed.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,229 @@
+      this.rooms[roomID] = room;
+    }
+
+    // Create a new peer with the given room and peer ID.
```
Yeah, great point! I think before this was doing something fairly complex so I moved it into a function and then forgot to delete this comment. But yeah, agree that in this case the comment is super obvious and not useful. I appreciate you sharing mental framework of comments needing to be above or below levels of abstraction, and will do a pass on my code to remove unneeded comments.

shalins commented on src/network/utils.ts:
```diff
@@ -0,0 +1,86 @@
+}
+
+/**
+ * Retreives an object from local storage that was saved from this
```
Yup, fair point, and I appreciate the level of detail here. I think I'm storing super small amounts of data here, bytes at most. But good to keep in mind when thinking through ways of persisting larger amounts of data.

kanyesthaker commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+   */
+  _handleWebsocketReceive(event: any) {
+    const receivedMetadata = parse(event.data);
+    const type = receivedMetadata.type;
```
just as an fyi, type is a reserved keyword in [strict mode](https://www.tektutorialshub.com/typescript/identifiers-keywords-in-typescript/#strict-mode-reserved-words)!

shalins replied:
Good to know!

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+    // Delete a peer by `filter`ing instead of using `splice` with
+    // indices, since we could run into races if multiple clients
+    // are joining and leaving at the same time.
+    this.rooms[roomID].peers = peersInRoom.filter(
```
So couple things here. The case you mentioned is basically an example of why you want thread-safe data structures, through some model (such as locks, actor model, etc.) that makes it safe for processes or threads to access things concurrently. In our particular scenario, we are using a Node.JS server, which is single threaded, and maintains a queue of requests so that things only happen one after another, preventing the sort of race you highlighted above.

That being said, it's still not a good idea to index directly into these data structure since we could run into weird issues where one peer is in the middle of disconnecting and another peer joins, and the new peer is added to the data structure, changing the position of the elements and causing us to run into an index out of bounds, which in javascript just returns `undefined`. 

Though this code is not thread safe, it does make it so the specific order of the elements do not matter in the data structure, prevent a different class of bugs.

Note: even though Node is single threaded, you can still have race conditions, for example if you're querying into a database asynchronously, etc.

shalins commented on signaler/types.ts:
```diff
@@ -0,0 +1,17 @@
+    ws?: WebSocket;
+}
+
+export interface Room {
```
It's not used in any of the logic, but I agree that semantically it should indeed have a `roomID` property. Added,

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+     * @param {WebSocket} ws WebSocket instance that disconnected.
+     */
+  handleClose(ws: WebSocket) {
+    // Find the peerID and roomID from the websocket instance.
```
Some recent changes makes this particular functionality only 2 lines of code, and it's not reused anywhere else. So for now will just keep in this function.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+
+    // If the connection with this PeerID is not found or undefined,
+    // return early.
+    if (!this.connections[peerID]) {
```
Arrays are a bit weird in that JS would just return `undefined` if this was not found. In either case, this is fixed by using a `Map`.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+    );
+
+    // Remove the connection.
+    delete this.connections[peerID];
```
Using a `Map` now.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+    // Check if the room exists. If not, create it.
+    if (!(roomID in this.rooms)) {
+      const room: Room = {peers: []};
+      this.rooms[roomID] = room;
```
Using a `Map` now.

shalins commented on signaler/types.ts:
```diff
@@ -0,0 +1,17 @@
+}
+
+export interface Room {
+    peers: Peer[];
```
Using a `Map` now.

shalins replied:
Also, yes it should be a list of `Peer`s, where `Peer` _is_ an object like you mentioned -- it's defined right above the `export interface Room` line.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+
+    // Cache the websocket object this peer is connected over so we can
+    // handle disconnect properly.
+    const connection: Connection = {peer: peer, ws: ws};
```
Yup, changed it to a map, albeit from `ws: Connection`, since that ended up being the most convenient.

shalins commented on signaler/server.ts:
```diff
@@ -0,0 +1,227 @@
+      peer = {
+        roomID: roomID,
+        peerID: peerID ? peerID : uuid().toString(),
+        initiator: !initiatorsPresent,
```
Yup, moved initiator change logic to the disconnect function. Good point.

shalins commented on src/network/communicator.ts:
```diff
@@ -0,0 +1,75 @@
+   * @param {boolean} debug Whether or not to enable debug logging.
+   */
+  constructor(roomID: string, debug = false) {
+    log.setLevel(debug ? log.levels.DEBUG : log.levels.SILENT);
```
Actually I'll have to agree with Kanyes on this one. He's right that log level should not be set in the constructor of a library. There are two ways to fix this, one is by setting the log level in the main `index.tsx` file (using `log.setLevel(xxxxxx)`) and then changing that level whenever we want to allow for more or less verbose logging. However the better way is to simply read from the environment whenever the program executes, so that changing the level of verbosity can happen without editing the code. So basically, in the main project directory, run `REACT_APP_LOG_LEVEL=xxxxxxx npm start` to start with a specific [log level](https://www.npmjs.com/package/loglevel), otherwise default is `warn`. Note that the variable is named such because React apps require environment variables to start with `REACT_APP_` to avoid accidentally reading private env vars named the same thing on your system.

shalins commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,299 @@
+     * @param {any} data The data that was sent with the connection event.
+     */
+  _handlePeerConnection(data: any) {
+    this.isPeerConnected = true;
```
Good catch, fixed.

shalins commented on signaler/README.md:
```diff
@@ -0,0 +1,122 @@
+
+function  Peer() {
+	const  location = useLocation();
+	const  communicator = useRef<Communicator>();
```
Done! This was a great idea to simplify the API, so I've added the corresponding hook and vastly simplified how the class can be used in the React code. The README has been updated to reflect this change.

shalins commented:
Done with first pass! I've basically addressed all comments except adding tests and fixing some of the `any` types to more descriptive types, since those are somewhat larger changes and this PR is already giant in scope. When you guys get a chance, please give this a second pass. Excited to get this merged!

samjgorman commented on signaler/README.md:
````diff
@@ -0,0 +1,155 @@
+The abstraction requires each peer to join a "room", which is essentially a collection of peers that should all be connected together via the signaling server and eventually over WebRTC. They should be able to communicate with each other via Communicator. Thus, in order to use Communicator, each peer must provide the RoomID, which is a string representing the room they wish to join.
+
+```
+const communicator = useCommunicator("test_room_1");
````
🔥

samjgorman commented on signaler/README.md:
```diff
@@ -0,0 +1,155 @@
+		event.preventDefault();
+		const  data = (event.currentTarget.elements[0] as HTMLInputElement).value;
+		if (communicator.current) {
+			communicator.current.send(data.toString());
```
Thanks for the descriptive documentation here-- very helpful

samjgorman commented on signaler/server.ts:
```diff
@@ -0,0 +1,236 @@
+    } else {
+      // Otherwise, make one of the peers in the room the initiator.
+      // Indexing at 0 is fine since we know there is at least one peer.
+      let peer = peersInRoom[0];
```
Default to const if not changing the value of the variable you're using

shalins replied:
Good catch -- done!

samjgorman commented on signaler/server.ts:
```diff
@@ -0,0 +1,236 @@
+    } else {
+      peer = {
+        roomID: roomID,
+        peerID: peerID ? peerID : uuid().toString(),
```
🚀

samjgorman commented on signaler/server.ts:
```diff
@@ -0,0 +1,236 @@
+      peersInRoom.forEach((p) => {
+        if (p.ws.readyState === WebSocket.OPEN) {
+          const initiate = createPayload(PayloadType.INITIATE, {});
+          p.ws.send(initiate);
```
Nonblocking: I prefer to avoid single letter variables unless using `i, j ` in loops or `e` for events

shalins replied:
Yeah I agree, it used to be `peer` but ESLint kept getting upset so changed it to `p` but agree it should be more descriptive. Fixed.

samjgorman commented on src/network/hooks/useCommunicator.ts:
```diff
@@ -0,0 +1,46 @@
+ * @param {CommunicatorCallback} onError - Callback when there is an error with the Communicator.
+ * @returns {Communicator} A reference to the Communicator.
+ */
+function useCommunicator(
```
Great work here! Can you change this to ES6 syntax for consistency?

ex)

const useCommunicator = () => { }

Also, happy to collaborate post-this PR to simplify the interface a bit. Essentially, we can likely rewrite so that a caller doesn't need to pass in these optional functions, but the hook returns these functions for use by the caller

shalins replied:
Yup, changed the syntax. And yes, that's a good idea. I'm down to simplify the interface more in a later PR.

shalins replied:
Btw, do you have an example in mind for your idea to return the functions by the hook (instead of the caller passing it in?) I'm not sure if I fully understand what you mean by that, so if you could share an example where it's following that design pattern, I'd be happy to incorporate it. Once I know what the right way to do it is, then I'm not sure it would take too long to include it. Let me know what you think. 

In my head, the reason why the caller needs to pass a function into the hook is because the caller may want to run an arbitrary block of code when events are fired. For example, the caller (for e.g. Player) might want to send a whole bunch of metadata when the `onConnect` event is fired.

samjgorman commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,316 @@
+    this.peerManager.roomID = peer.roomID;
+    this.peerManager.isInitiator = peer.initiator;
+
+    log.debug(
```
Nice work making this descriptive!

samjgorman reviewed (approved):
Looking great to me-- really excited to have this in the codebase. Thanks for your hard work incorporating reviews!

kanyesthaker commented on signaler/server.ts:
```diff
@@ -0,0 +1,246 @@
+   * @param {WebSocket.RawData} message Message received from the client.
+   */
+  handleReceive(ws: WebSocket, message: WebSocket.RawData) {
+    const parsed = JSON.parse(message.toString());
```
[optional nit] `parsedMessage` for clarity

shalins replied:
Done!

kanyesthaker commented on signaler/server.ts:
```diff
@@ -0,0 +1,246 @@
+   */
+  handleReceive(ws: WebSocket, message: WebSocket.RawData) {
+    const parsed = JSON.parse(message.toString());
+    const [type, data, peerID, roomID] = [
```
would prefer `messageType`, for reasons previously mentioned

shalins replied:
Ah, corrected it in Communicator class but missed it in the server code.

kanyesthaker commented on signaler/server.ts:
```diff
@@ -0,0 +1,246 @@
+      this.rooms.delete(roomID);
+      log.debug(`No other peers in the room, room ${roomID} deleted.`);
+    } else {
+      // Otherwise, make one of the peers in the room the initiator.
```
optional nit: stick w/ a consistent format for inline comments (`//` vs `/** */`)

shalins replied:
Missed this, nice catch!

kanyesthaker commented on src/network/signalingClient.ts:
```diff
@@ -0,0 +1,316 @@
+   * @param {any} data The data that was sent with the close event.
+   */
+  _handlePeerClose(data: any) {
+    this.isPeerConnected = false;
```
nice

kanyesthaker commented on src/network/utils.ts:
```diff
@@ -0,0 +1,86 @@
+/* eslint-disable no-unused-vars */
```
do we need this eslint rule?

shalins replied:
Nope, removed from here too.

kanyesthaker reviewed (approved):
lgtm, left a few optional nits, great work!

shalins commented:
Fixed everything based on your comments! Thanks again for the detailed and in-depth review!
